File: app/src/main/java/com/kakeibo/export/ExportActivity.java
package com.kakeibo.export;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.InterstitialAd;
import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.initialization.InitializationStatus;
import com.google.android.gms.ads.initialization.OnInitializationCompleteListener;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.Scope;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.api.client.extensions.android.http.AndroidHttp;
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.services.drive.Drive;
import com.google.api.services.drive.DriveScopes;
import com.google.api.services.drive.model.File;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;
import com.kakeibo.BuildConfig;
import com.kakeibo.KkbApplication;
import com.kakeibo.R;
import com.kakeibo.TabFragment2;
import com.kakeibo.util.UtilDate;
import com.kakeibo.util.UtilFiles;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import java.io.IOException;
import java.io.OutputStream;
import java.util.Collections;

import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

public class ExportActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    public static final String FILE_ORDER_DATE = "tmp_order_date.csv";
    public static final String FILE_ORDER_CATEGORY = "tmp_order_cat.csv";

    private static final int REQUEST_CODE_SIGN_IN = 1;
    private static final int REQUEST_CODE_OPEN_DOCUMENT = 2;

    private DriveServiceHelper mDriveServiceHelper;
    private String mOpenFileId;

    private Context _context;
    private static int mReportType;
    private GoogleSignInClient mGoogleSignInClient;
    private FirebaseAuth mFirebaseAuth;
    private String mStrDateFormat;

    private InterstitialAd mInterstitialAd;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_export);
        _context = this;

        mReportType = getIntent().getIntExtra("REPORT_VIEW_TYPE", 0);

        int dateFormat = KkbApplication.getDateFormat(R.string.pref_key_date_format);
        switch (dateFormat) {
            case 1: // MDY
                mStrDateFormat = UtilDate.getTodaysDate(UtilDate.DATE_FORMAT_MDY) + " kk:mm:ss";
                break;
            case 2: // DMY
                mStrDateFormat = UtilDate.getTodaysDate(UtilDate.DATE_FORMAT_DMY) + " kk:mm:ss";
                break;
            default:  // YMD
                mStrDateFormat = UtilDate.getTodaysDate(UtilDate.DATE_FORMAT_YMD) + " kk:mm:ss";
        }

        requestSignIn();

        /*** ads ***/
        MobileAds.initialize(this, new OnInitializationCompleteListener() {
            @Override
            public void onInitializationComplete(InitializationStatus initializationStatus) {}
        });
        mInterstitialAd = new InterstitialAd(this);
        if (BuildConfig.DEBUG) {
            mInterstitialAd.setAdUnitId("ca-app-pub-3940256099942544/1033173712");/*** in debug mode ***/
        } else {
            mInterstitialAd.setAdUnitId(getString(R.string.main_banner_ad));
        }
        mInterstitialAd.loadAd(new AdRequest.Builder().build());
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent resultData) {
        switch (requestCode) {
            case REQUEST_CODE_SIGN_IN:
                if (resultCode == Activity.RESULT_OK && resultData != null) {
                    handleSignInResult(resultData);
                }
                break;

            case REQUEST_CODE_OPEN_DOCUMENT:
                if (resultCode == Activity.RESULT_OK && resultData != null) {
                    Uri uri = resultData.getData();
                    if (uri != null) {
//                        openFileFromFilePicker(uri);
                        try {
                            OutputStream os = getContentResolver().openOutputStream(uri);
                            if( os != null ) {

                                String content = "";
                                if (mReportType == TabFragment2.REPORT_BY_DATE) {
                                    content = UtilFiles.getFileValue(FILE_ORDER_DATE, _context);
                                } else if (mReportType == TabFragment2.REPORT_BY_CATEGORY) {
                                    content = UtilFiles.getFileValue(FILE_ORDER_CATEGORY, _context);
                                } else {
                                    content = "Empty Report";
                                }

                                os.write(content.getBytes());
                                os.close();

                                /*** ads ***/
                                if (mInterstitialAd.isLoaded()) {
                                    mInterstitialAd.show();
                                } else {
                                    Log.d("TAG", "The interstitial wasn't loaded yet.");
                                }
                            }
                        }
                        catch(IOException e) {
                            e.printStackTrace();
                        }
                    }
                }
                finish();
                break;
        }

        super.onActivityResult(requestCode, resultCode, resultData);
    }

    /**
     * Starts a sign-in activity using {@link #REQUEST_CODE_SIGN_IN}.
     */
    private void requestSignIn() {
        Log.d(TAG, "Requesting sign-in");

        GoogleSignInOptions signInOptions =
                new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                        .requestIdToken(getString(R.string.default_web_client_id))
                        .requestEmail()
                        .requestScopes(new Scope(DriveScopes.DRIVE_FILE))
                        .build();
        mGoogleSignInClient = GoogleSignIn.getClient(this, signInOptions);
        mFirebaseAuth = FirebaseAuth.getInstance();
        startActivityForResult(mGoogleSignInClient.getSignInIntent(), REQUEST_CODE_SIGN_IN);
    }

    private void requestSignOut() {
        /*** signout from firebase ***/
        if (mFirebaseAuth != null) mFirebaseAuth.signOut();

        /*** signout from googleSignInClient ***/
        if (mGoogleSignInClient != null) {
            mGoogleSignInClient.signOut()
                    .addOnCompleteListener(this, new OnCompleteListener<Void>() {
                        @Override
                        public void onComplete(@NonNull Task<Void> task) { }
                    });
        }
    }

    /**
     * Handles the {@code result} of a completed sign-in activity initiated from {@link
     * #requestSignIn()}.
     */
    private void handleSignInResult(Intent result) {
        GoogleSignIn.getSignedInAccountFromIntent(result)
                .addOnSuccessListener(googleAccount -> {
                    Log.d(TAG, "Signed in as " + googleAccount.getEmail());

                    // Use the authenticated account to sign in to the Drive service.
                    GoogleAccountCredential credential =
                            GoogleAccountCredential.usingOAuth2(
                                    this, Collections.singleton(DriveScopes.DRIVE_FILE));
                    credential.setSelectedAccount(googleAccount.getAccount());
                    Drive googleDriveService =
                            new Drive.Builder(
                                    AndroidHttp.newCompatibleTransport(),
                                    new GsonFactory(),
                                    credential)
                                    .setApplicationName("Drive API Migration")
                                    .build();

                    // The DriveServiceHelper encapsulates all REST API and SAF functionality.
                    // Its instantiation is required before handling any onClick actions.
                    mDriveServiceHelper = new DriveServiceHelper(googleDriveService);

                    /*** firebase login ***/
                    AuthCredential authCredential = GoogleAuthProvider.getCredential(googleAccount.getIdToken(), null);
                    mFirebaseAuth.signInWithCredential(authCredential)
                            .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                                @Override
                                public void onComplete(@NonNull Task<AuthResult> task) {
                                    if (task.isSuccessful()) {
                                        // Sign in success, update UI with the signed-in user's information
                                        FirebaseUser user = mFirebaseAuth.getCurrentUser();
                                        if (user!=null) Log.d(TAG, "Firebase user.getDisplayName()= "+ user.getDisplayName());
                                    } else {
                                        Log.w(TAG, "signInWithCredential:failure", task.getException());
                                        finish();
                                    }
                                }
                            });

                    openFilePicker();
                })
                .addOnFailureListener(exception -> Log.e(TAG, "Unable to sign in.", exception));
    }

    /**
     * Opens the Storage Access Framework file picker using {@link #REQUEST_CODE_OPEN_DOCUMENT}.
     */
    private void openFilePicker() {
        if (mDriveServiceHelper != null) {
            Log.d(TAG, "Opening file picker.");

            Intent pickerIntent = mDriveServiceHelper.createFilePickerIntent();

            // The result of the SAF Intent is handled in onActivityResult.
            startActivityForResult(pickerIntent, REQUEST_CODE_OPEN_DOCUMENT);
        }
    }

    /**
     * Opens a file from its {@code uri} returned from the Storage Access Framework file picker
     * initiated by {@link #openFilePicker()}.
     */
    private void openFileFromFilePicker(Uri uri) {
        if (mDriveServiceHelper != null) {
            Log.d(TAG, "Opening " + uri.getPath());

            mDriveServiceHelper.openFileUsingStorageAccessFramework(getContentResolver(), uri)
                    .addOnSuccessListener(nameAndContent -> {
                        setReadOnlyMode();
                        createFile();
                    })
                    .addOnFailureListener(exception ->
                            Log.e(TAG, "Unable to open file from picker.", exception));
        }
    }

    /**
     * Creates a new file via the Drive REST API.
     */
    private void createFile() {
        if (mDriveServiceHelper != null) {
            Log.d(TAG, "Creating a file.");

            mDriveServiceHelper.createFile(UtilDate.getTodaysDate(mStrDateFormat), mReportType, _context)
                    .addOnSuccessListener(fileId -> {
//                        readFile(fileId);
                        showMessage(getString(R.string.file_created));

                        /*** ads ***/
                        if (mInterstitialAd.isLoaded()) {
                            mInterstitialAd.show();
                        } else {
                            Log.d("TAG", "The interstitial wasn't loaded yet.");
                        }
                        finish();
                    })
                    .addOnFailureListener(exception -> {
                        Log.e(TAG, "Couldn't create file.", exception);
                        finish();
                    });
        }
    }

    /**
     * Retrieves the title and content of a file identified by {@code fileId} and populates the UI.
     */
    private void readFile(String fileId) {
        if (mDriveServiceHelper != null) {
            Log.d(TAG, "Reading file " + fileId);

            mDriveServiceHelper.readFile(fileId)
                    .addOnSuccessListener(nameAndContent -> {
                        setReadWriteMode(fileId);
                    })
                    .addOnFailureListener(exception ->
                            Log.e(TAG, "Couldn't read file.", exception));
        }
    }

    /**
     * Saves the currently opened file created via {@link #createFile()} if one exists.
     */
    private void saveFile() {
        if (mDriveServiceHelper != null && mOpenFileId != null) {
            String fileName = "fileName";
            String fileContent = "fileContent";
            mDriveServiceHelper.saveFile(mOpenFileId, fileName, fileContent)
            .addOnFailureListener(exception ->
                    Log.e(TAG, "Unable to save file via REST.", exception));
        }
    }

    /**
     * Queries the Drive REST API for files visible to this app and lists them in the content view.
     */
    private void query() {
        if (mDriveServiceHelper != null) {
            mDriveServiceHelper.queryFiles()
                    .addOnSuccessListener(fileList -> {
                        StringBuilder builder = new StringBuilder();
                        for (File file : fileList.getFiles()) {
                            builder.append(file.getName()).append("\n");
                        }
                        String fileNames = builder.toString();
                        // todo should be disposable
//                        mFileTitleEditText.setText("File List");
//                        mDocContentEditText.setText(fileNames);

                        setReadOnlyMode();
                    })
                    .addOnFailureListener(exception -> Log.e(TAG, "Unable to query files.", exception));
        }
    }

    /**
     * Updates the UI to read-only mode.
     */
    private void setReadOnlyMode() {
        mOpenFileId = null;
    }

    /**
     * Updates the UI to read/write mode on the document identified by {@code fileId}.
     */
    private void setReadWriteMode(String fileId) {
        mOpenFileId = fileId;
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.d(TAG, "onStop() called");
        requestSignOut();
    }

    protected void showMessage(String message) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
    }

    public void screenTapped(View view) { finish(); }
}

File: app/src/main/java/com/kakeibo/ui/ExportActivity.kt
package com.kakeibo.ui

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.Scope
import com.google.android.gms.tasks.Task
import com.google.api.client.extensions.android.http.AndroidHttp
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.json.gson.GsonFactory
import com.google.api.services.drive.Drive
import com.google.api.services.drive.DriveScopes
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.kakeibo.BuildConfig
import com.kakeibo.R
import com.kakeibo.ui.export.DriveServiceHelper
import com.kakeibo.ui.model.Medium
import com.kakeibo.ui.viewmodel.FirebaseUserViewModel
import com.kakeibo.util.UtilFiles
import java.io.IOException

class ExportActivity : AppCompatActivity() {

    companion object {
        private const val TAG = "ExportActivity"
        const val FILE_ORDER_DATE = "tmp_order_date.csv"
        const val FILE_ORDER_CATEGORY = "tmp_order_cat.csv"
        private var mReportType = 0
    }

    private var mDriveServiceHelper: DriveServiceHelper? = null
    private var mGoogleSignInClient: GoogleSignInClient? = null
    private var mFirebaseAuth: FirebaseAuth? = null
    private var _interstitialAd: InterstitialAd? = null

    private val _authenticationViewModel: FirebaseUserViewModel by viewModels()

    val _requestSignIn = registerForActivityResult (
        ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == RESULT_OK) {
            val task: Task<GoogleSignInAccount> = GoogleSignIn.getSignedInAccountFromIntent(result.data)
            handleSignInResult(task)
        }
    }
    val _requestOpenDocument = registerForActivityResult (
        ActivityResultContracts.StartActivityForResult()
    ) {result: ActivityResult ->
        if (result.resultCode == RESULT_OK) {
            result.data?.data?.let {
                try {
                    val os = contentResolver.openOutputStream(it)
                    if (os != null) {
                        val content: String? = when (mReportType) {
                            Medium.FRAGMENT_REPORT_DATE_MONTHLY -> {
                                UtilFiles.getFileValue(FILE_ORDER_DATE, this)
                            }
                            Medium.FRAGMENT_REPORT_CATEGORY_MONTHLY -> {
                                UtilFiles.getFileValue(FILE_ORDER_CATEGORY, this)
                            }
                            else -> {
                                "Empty Report"
                            }
                        }
                        os.write(content!!.toByteArray())
                        os.close()

                        /* ads */
                        if (_interstitialAd != null) {
                            _interstitialAd!!.show(this)
                        } else {
                            Log.d("TAG", "The interstitial wasn't loaded yet.")
                        }

                        Toast.makeText(this, R.string.file_created, Toast.LENGTH_LONG).show()
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
        finish()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_export)
        mReportType = intent.getIntExtra("REPORT_VIEW_TYPE", 0)

        _authenticationViewModel.firebaseUser.observe(this, {
            if (it == null) {
                requestSignIn()
            } else {
                val googleAccount = GoogleSignIn.getLastSignedInAccount(this)
                googleAccount?.let {
                    connectToGDrive(googleAccount)
                }
            }
        })

        /* ads */
        val adRequest = AdRequest.Builder().build()
        var adUnitId = getString(R.string.upload_ad)
        if (BuildConfig.DEBUG) { /* in debug mode */
            adUnitId = "ca-app-pub-3940256099942544/1033173712"
        }
        InterstitialAd.load(this, adUnitId, adRequest, object : InterstitialAdLoadCallback() {
            override fun onAdLoaded(interstitialAd: InterstitialAd) {
                _interstitialAd = interstitialAd
            }

            override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                _interstitialAd = null
            }
        })
    }

    /*
     * Starts a sign-in activity using [.REQUEST_CODE_SIGN_IN].
     */
    private fun requestSignIn() {
        Log.d(TAG, "Requesting sign-in")
        val signInOptions = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .requestScopes(Scope(DriveScopes.DRIVE_FILE))
                .build()
        mGoogleSignInClient = GoogleSignIn.getClient(this, signInOptions)
        mFirebaseAuth = FirebaseAuth.getInstance()
        mGoogleSignInClient?.let {
            _requestSignIn.launch(it.signInIntent)
        }
    }

    /*
     * Handles the `result` of a completed sign-in activity initiated from [ ][.requestSignIn].
     */
    private fun handleSignInResult(completedTask: Task<GoogleSignInAccount>) {
        try {
            val googleAccount = completedTask.getResult(ApiException::class.java)
            googleAccount?.let {
                connectToGDrive(googleAccount)
            }
        } catch (e: ApiException) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.e(TAG, "signInResult:failed code=" + e.statusCode)
            Toast.makeText(this, "Google Sign-in failed. Error: $e", Toast.LENGTH_LONG).show()
            finish()
        }
    }

    private fun connectToGDrive(googleAccount: GoogleSignInAccount) {
        // Signed in successfully, show authenticated UI.
        // Use the authenticated account to sign in to the Drive service.
        val credential = GoogleAccountCredential.usingOAuth2(this, setOf(DriveScopes.DRIVE_FILE))
        credential.selectedAccount = googleAccount.account
        val googleDriveService = Drive.Builder(
                AndroidHttp.newCompatibleTransport(),
                GsonFactory(),
                credential)
            .setApplicationName("Drive API Migration")
            .build()

        // The DriveServiceHelper encapsulates all REST API and SAF functionality.
        // Its instantiation is required before handling any onClick actions.
        mDriveServiceHelper = DriveServiceHelper(googleDriveService)
        /* firebase login  */
        val authCredential = GoogleAuthProvider.getCredential(googleAccount.idToken, null)
        mFirebaseAuth?.let {
            it.signInWithCredential(authCredential).addOnCompleteListener(this) { task ->
                if (task.isSuccessful) {
                    // Sign in success, update UI with the signed-in user's information
                    val user = it.currentUser
                    if (user != null) Log.d(TAG, "Firebase user.getDisplayName()= " + user.displayName)

                    _authenticationViewModel.updateFirebaseUser()
                } else {
                    Log.w(TAG, "signInWithCredential:failure", task.exception)
                    Toast.makeText(this, "signInWithCredential:failure", Toast.LENGTH_LONG).show()
                    finish()
                }
            }
        }
        openFilePicker()
    }

    /*
     * Opens the Storage Access Framework file picker using [.REQUEST_CODE_OPEN_DOCUMENT].
     */
    private fun openFilePicker() {
        if (mDriveServiceHelper != null) {
            Log.d(TAG, "Opening file picker.")
            val pickerIntent = mDriveServiceHelper?.createFilePickerIntent()

            // The result of the SAF Intent is handled in onActivityResult.
            _requestOpenDocument.launch(pickerIntent)
        }
    }

    fun screenTapped(view: View?) {
        finish()
    }
}